### DISCUSSION TITLE
cq码解析

### DISCUSSION THREAD

**Discussion by snowy-yang** [nonebot/discussions#459] [问答] [answered]
nonebot2中我想要解析一段json的cq码，我应该怎么解析他，或者有什么办法获取xml或json格式的数据，我现在被cq码搞得心烦意乱，不知道cq码发明来干什么

**Comment by yanyongyu** [comment]
在nb里你不应该需要解析cq码，你是在哪里获取的cq码，请给出你的代码与日志

**Reply by snowy-yang** [reply]
从message里获取的

**Reply by snowy-yang** [reply]
明天给你代码

**Reply by snowy-yang** [reply]
我记得是依赖注入的Message获取到的

**Reply by snowy-yang** [reply]
> 在nb里你不应该需要解析cq码，你是在哪里获取的cq码，请给出你的代码与日志

希望给一个方法获取原内容

**Reply by yanyongyu** [reply]
在nb里你拿到的应该是`Message`对象而非cq码字符串，请检查你的代码

**Reply by snowy-yang** [reply]
那么message对象如何获取内容，我发现不能获取内容于是将它str了

**Reply by snowy-yang** [reply]
> 在nb里你拿到的应该是`Message`对象而非cq码字符串，请检查你的代码

看到回答哈

**Reply by yanyongyu** [reply, marked as answer]
请参考[消息处理文档](https://nonebot.dev/docs/tutorial/message)

**Reply by snowy-yang** [reply]
> 请参考[消息处理文档](https://nonebot.dev/docs/tutorial/message)

啊

**Comment by synodriver** [comment]
一般不是构建Message对象吗？或者你的意思是把pydantic的模型转成json？这种用dict或者model_dump方法

**Reply by synodriver** [reply]
还有model_validate等

**Reply by snowy-yang** [reply]
我是接受到一段cq码需要解析为dict

**Reply by snowy-yang** [reply]
> 还有model_validate等

这些是什么

**Reply by synodriver** [reply]
nonebot.utils有个DataclassEncoder，你可以参考下

**Reply by snowy-yang** [reply]
> nonebot.utils有个DataclassEncoder，你可以参考下

怎么个使用法

**Reply by synodriver** [reply]
json.dumps(data, cls=DataclassEncoder)这样

**Reply by snowy-yang** [reply]
> json.dumps(data, cls=DataclassEncoder)这样

明早试试，没见过，真是一个神奇的东西

**Reply by synodriver** [reply]
这东西就是把Message对象序列化的

**Reply by synodriver** [reply]
为json量身定制的，json默认不认你的自定义类

**Reply by snowy-yang** [reply]
> 为json量身定制的，json默认不认你的自定义类

都有这个了nonebot2为什么不自己序列化完了给我一个返回值

**Reply by snowy-yang** [reply]
> 为json量身定制的，json默认不认你的自定义类

还有cq码不是onebot11协议的吗，不应该从协议适配器里获取到的吗

**Reply by synodriver** [reply]
所有让你塞Message对象去序列化

**Reply by snowy-yang** [reply]
> 所有让你塞Message对象去序列化

明早试试

**Reply by synodriver** [reply]
你想得到[这个的结果](https://github.com/nonebot/adapter-onebot/blob/master/nonebot/adapters/onebot/v11/adapter.py#L134-L144) ？

**Reply by snowy-yang** [reply]
> 你想得到[这个的结果](https://github.com/nonebot/adapter-onebot/blob/master/nonebot/adapters/onebot/v11/adapter.py#L134-L144) ？

你有没有空试试https://nonebot.dev/docs/tutorial/message

**Reply by snowy-yang** [reply]
> 你想得到[这个的结果](https://github.com/nonebot/adapter-onebot/blob/master/nonebot/adapters/onebot/v11/adapter.py#L134-L144) ？

刚刚另一个大佬告诉我可以这样获取

**Reply by synodriver** [reply]
你不是要json吗
